### DISCUSSION TITLE
Am I supposed to be using `deno task` in order to run something something like `npm:create-svelte`?

### DISCUSSION THREAD

**Discussion by bzm3r** [denoland/deno#25744] [Q&A] [answered]
[Previous discussion here](https://github.com/denoland/deno/issues/25373#issuecomment-2336799994) but I got no feedback there, so I'm bringing it over here.

Apparently, `deno run` is going out the window for certain things, but this is confusing me as a Deno beginner:

> We've discussed this internally, and this is working as intended.
> 
> `deno run` always prioritizes running _code_, which means it first and foremost treats it's arguments as file paths. For example you also can not run a task named `./task.ts` with `deno run`, as it will attempt to load that file. Only if a task does not look like a file path or module specifier will we attempt to look for tasks and run them.
> 
> Instead you should use `deno task` to run tasks like this (or better, rename your task!).

Hi! Thank you for your explanation.

I'm new to `deno`, and it's not really clear to me how `deno task` fits here [because of how the docs explain `deno task`](https://docs.deno.com/runtime/reference/cli/task_runner/) (emphasis mine):

> `deno task` provides a cross-platform way to define and execute *custom commands specific to a codebase*.

Suppose I want to execute `npm:create-svelte@latest`. This is not a command specific to a codebase, and therefore "not a task"! It is a command meant to initialize a codebase. 

Suppose I were to generalize a bit, and define a codebase to be "my home directory". Then, I guess I would put a `deno.json` file there, with something like? (not sure)

```json
{
  "tasks": {
    "create-svelte": "deno run -A npm:create-svelte@latest"
  }
}
```

I am not sure about the above because setting up a `deno.json` just to do a `deno run` seems very circular...

...to the point where I did not expect it to work.

Yet it did! And when it did, I was surprised, and tried a bare `deno run -A npm:create-svelte@latest`....

...which also worked? Huh. I made sure I've enabled `DENO_FUTURE`:

```
> $env:DENO_FUTURE
1
```

So, why does this work now? I am using:

```
> deno --version
deno 1.46.3 (stable, release, x86_64-pc-windows-msvc)
v8 12.9.202.5-rusty
typescript 5.5.2
```

So...is `deno task` the recommended way to run something like `npm:create-svelte` or is `deno run` still the way to go?

**Comment by marvinhagemeister** [comment]
I think you're taking the `task` concept a bit too much to the extreme. If you want to fire a one-off command for `npm:create-svelte@latest`, then you should just do that by calling `deno run -A npm:create-svelte@latest`. Whilst you can also put this in a task, it would be kinda pointless.

The motivation behind tasks is to abstract common complex tasks that you need to run daily. A typical use case is that there is a `dev` task that spins up your app in development mode, sets up file watching and maybe does other work needed to launch your application. Let's use a made up example and pretend that this would be the command to launch your app in development mode:

```sh
deno run -A --watch-hmr my-app.ts
```

Now you could of course always type out the whole command by hand yourself. This gets very tedious though. You could put this in a shell file, but then have to deal with different developers using different shells and operating systems, which all have a different syntax. Windows has PowerShell, macOS has zsh and linux typically has bash or sometimes fish. It's not something you want to bother with as a developer, and neither should your team mates.

So what if you could just call `deno taks dev` instead? That way you don't need to remember the exact command for a project. Most projects have different commands for this and having to remember every one of them is annoying, especially if you're working in a team that works on many project at the same time.

We can solve this headache by adding a `dev` task to all projects. That way developers on the different projects don't need to remember whether this projects needs to copy some files first before launching, how to set up the file watcher for this particular project, what the entrypoint name for a certain project is, etc. The list of things to remember gets pretty long the more complex a project grows.

This all can be solved with tasks. Let's add a task for development setup of our imaginary project:

```json
{
  "tasks": {
    "dev": "deno run -A --watch-hmr my-app.ts"
  }
}
```

This means now you can just call `deno task dev` and don't have to remember anything else. In other words: The task feature is only there fore convenience.  It allows you to abstract away more complex commands in something that's easier to remember. In most projects you'll find a `build`, a `start` or a `dev` task respectively. As you discovered, they make no sense for one off things like bootstrapping a new project. That said, nothing prevents you from creating a task out of any `deno` command. You can literally put anything in a task. Whether that is a useful thing to do is a different question.

**Reply by bzm3r** [reply]
@marvinhagemeister Thank you for your answer, but I think I did not write my question well because of my confusion, so maybe the following will shed some light:

- I appreciate your refresher on tasks because I was indeed worried I was misunderstanding tasks, and now I think what you wrote confirms for me that conceptually `deno task` is not what I should be using in order to run `npm:create-svelte@latest` (because a task is something *we define* as a "macro" within the context of a project)
  - if you are not in the context of a Deno project, then it does not make sense to define a task in order to run `npm:create-svelte@latest`; you just want to run `deno run -A npm:create-svelte@latest` once (it is a simple, one-off command)! 
- I fell into this confusion because, about two weeks ago I ran into the issue where: `deno run -A npm:create-svelte@latest` was throwing an error:
```
error: Could not find a matching package for 'npm:create-svelte@latest' in a package.json file. You must specify this as a package.json dependency when the node_modules folder is not managed by Deno.
```
- Apparently, according to [someone who responded on Discord](https://discord.com/channels/684898665143206084/684898665151594506/1280295285385461903), this was due to https://github.com/denoland/deno/issues/25232 and they created for me the following issue: https://github.com/denoland/deno/issues/25373
  - I am not entirely sure yet, but I think they might have misunderstood, and I should not have let them create the issue and instead created one myself? 
- In any case, the discussion in that confused me. It seemed to me that instead of `deno run -A npm:create-svelte@latest` (as I would usually do) I was supposed to use instead use `deno task`??? I found it difficult to understand lucacasonato's [comment](https://github.com/denoland/deno/issues/25373#issuecomment-2326110544):

> We've discussed this internally, and this is working as intended.

> deno run always prioritizes running code, which means it first and foremost treats it's arguments as file paths. For example you also can not run a task named ./task.ts with deno run, as it will attempt to load that file. Only if a task does not look like a file path or module specifier will we attempt to look for tasks and run them.

> Instead you should use deno task to run tasks like this (or better, rename your task!).

- the point that confused me was this last sentence "Instead you should use deno task to run tasks like this, or better yet, rename your task"...
  - and I think this is because the issue author thought my problem was coming from trying to run `deno task` with `npm:create-svelte@latest`? 
  - I am not sure, but there is some miscommunication somewhere


In any case, what would help me right now is to get some clarity on:
  - why is it that I was getting that error when running `deno run -A npm:create-svelte@latest`? 
  - why is it that I am no longer getting that error? Can I just go on with using `deno run -A npm:create-svelte@latest` as if none of the intervening stuff happened?

**Reply by marvinhagemeister** [reply, marked as answer]
> why is it that I was getting that error when running `deno run -A npm:create-svelte@latest`?

That was a bug that we fixed. There shouldn't be any error.

> why is it that I am no longer getting that error?

Because we fixed that bug. It occurred because we also added support for running `tasks` defined in `deno.json` via `deno run` to make it easier for folks swichting from `npm`. We missed an important edge case leading to the weird behavior you described.

> Can I just go on with using `deno run -A npm:create-svelte@latest` as if none of the intervening stuff happened?

Yes, please do. That's the way to do it 👍

**Reply by bzm3r** [reply]
Thank you so much for taking out the time to explain things to me!
